libdrgn/kdump.c

```diff
@@ -126,9 +126,72 @@ struct drgn_error *drgn_program_set_kdump(struct drgn_program *prog)
 
 	prog->flags |= DRGN_PROGRAM_IS_LINUX_KERNEL;
 	drgn_program_set_platform(prog, &platform);
+	prog->kdump_ctx = ctx;
 	return NULL;
 
 err:
 	kdump_free(ctx);
 	return err;
 }
+
+struct drgn_error *drgn_program_prstatus_kdump_iter(struct drgn_program *prog,
+						     drgn_prstatus_callback_fn cb_fn)
+{
+	kdump_num_t ncpus;
+	kdump_status ks;
+
+	ks = kdump_get_number_attr(prog->kdump_ctx, "cpu.number", &ncpus);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+					 "kdump_get_number_attr(cpu.number): %s",
+					 kdump_get_err(prog->kdump_ctx));
+	}
+
+	for (int i = 0; i < ncpus; i++) {
+		char *attr_name = NULL;
+		kdump_attr_ref_t prstatus_ref;
+		kdump_attr_t prstatus_attr;
+		unsigned char *prstatus_data;
+		size_t prstatus_size;
+		struct drgn_prstatus_map_entry entry;
+		struct drgn_error *err = NULL;
+		int ret;
+
+		ret = asprintf(&attr_name, "cpu.%d.PRSTATUS", i);
+		if (ret == -1)
+			return &drgn_enomem;
```

Comment:
This can be a statically-sized buffer.

```
$ python3 -c 'print(len("cpu.%d.PRSTATUS" % (2**32 - 1)))'
23
```

`char attr_name[24];` should suffice.